components/calendar/src/persian.rs

```diff
@@ -0,0 +1,344 @@
+        date.0.offset_date(offset)
+    }
+
+    #[allow(clippy::field_reassign_with_default)]
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        _calendar2: &Self,
+        _largest_unit: DateDurationUnit,
+        _smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        date1.0.until(date2.0, _largest_unit, _smallest_unit)
+    }
+
+    fn year(&self, date: &Self::DateInner) -> types::FormattableYear {
+        Self::year_as_persian(date.0.year)
+    }
+
+    fn month(&self, date: &Self::DateInner) -> types::FormattableMonth {
+        date.0.solar_month()
+    }
+
+    fn day_of_month(&self, date: &Self::DateInner) -> types::DayOfMonth {
+        date.0.day_of_month()
+    }
+
+    fn day_of_year_info(&self, date: &Self::DateInner) -> types::DayOfYearInfo {
+        let prev_year = date.0.year - 1;
+        let next_year = date.0.year + 1;
+        types::DayOfYearInfo {
+            day_of_year: date.0.day_of_year(),
+            days_in_year: date.0.days_in_year(),
+            prev_year: Persian::year_as_persian(prev_year),
+            days_in_prev_year: Persian::days_in_year_direct(prev_year),
+            next_year: Persian::year_as_persian(next_year),
+        }
+    }
+
+    fn debug_name(&self) -> &'static str {
+        "Persian"
+    }
+    // Missing any_calendar persian tests, the rest is completed
+    fn any_calendar_kind(&self) -> Option<AnyCalendarKind> {
+        Some(AnyCalendarKind::Persian)
+    }
+}
+
+impl Persian {
+    pub fn new() -> Self {
+        Self
+    }
+
+    fn fixed_from_arithmetic_persian(date: PersianDateInner) -> RataDie {
+        let month = date.0.month;
+
+        let day = date.0.day;
+
+        let mut year: i32 = if 0 < date.0.year {
+            date.0.year - 474
+        } else {
+            date.0.year - 473
+        };
+        year = (year % 2820) + 474;
+
+        let result: i32 = 1_948_320 - 1
+            + (1029983 * (year / 2820))
+            + (365 * (year - 1))
+            + ((31 * year - 5) / 128)
+            + if month <= 7 {
+                31 * (month as i32 - 1)
+            } else {
+                30 * (month as i32 - 1) + 6
+            }
+            + day as i32;
+
+        RataDie::new(result as i64)
+    }
+
+    fn arithmetic_persian_from_fixed(date: RataDie) -> Date<Persian> {
+        let _year = Self::arithmetic_persian_year_from_fixed(date);
+        let day_of_year = (date
+            - Self::fixed_from_arithmetic_persian(PersianDateInner(ArithmeticDate {
+                year: _year,
+                month: 1,
+                day: 1,
+                marker: (core::marker::PhantomData),
+            })))
+            + 1;
+        let _month = if day_of_year <= 186 {
+            libm::ceil(day_of_year as f64 / 31.0) as u8
+        } else {
+            libm::ceil((day_of_year - 6) as f64 / 30.0) as u8
```

Comment:
You can do this in integer space without pulling in `libm`:
```suggestion
            ((day_of_year + 30) / 31) as u8
        } else {
            ((day_of_year - 6 + 29) / 30) as u8
```

Reply:
done! Thanks for the suggestion

Reply:
I think @robertbastian's code works for positive numbers only, which I think is probably what we always have here, but the book specifically requests a ceiling division for this operation. Since that seems like a useful operation to generalize, we implemented a `ceil_div` function, which also gets the job done. Both are way better than `libm::ceil`